include/deal.II/fe/fe_enriched.h

```diff
@@ -691,7 +696,502 @@ class FE_Enriched : public FiniteElement<dim, spacedim>
+     * using the function color_predicates.
+     *
+     * Each active FE index has a set of colors associated with it.
+     * A cell with an active FE index i has a set of colors given by
+     * <code>fe_sets[i]</code>. An active FE index with color {a,b}
+     * means that the cell has two active predicates (i.e they return true
+     * for the cell) of color a and b.
+     *
+     * Eg: fe_sets = { {}, {1}, {2}, {1,2} } means
+     * Cells with active FE index 0 have no predicates associated.
+     * Cells with index 1 have a active predicate with color 1.
+     * Cells with index 2 have a active predicate with color 2.
+     * Cells with index 3 have active predicates with color 1 and color 2.
+     *
+     * A map of maps cellwise_color_predicate_map is used to associate
+     * predicate colors in cells with predicate ids. For this purpose, each
+     * cell is given a unique id which is stored in material id for now.
+     * When the grid is refined, material id is inherited to the children, so
+     * map which associates material id with color map will still be relevant.
+     *
+     * Now the color map can be explained with an example. If the cell with
+     * material id 100 has active predicates 4 (color = 1) and 5 (color = 2),
+     * the map will insert pairs (1, 4) and (2, 5) at key 100 (i.e unique id
+     * of cell is mapped with a map which associates color with predicate id).
+     *
+     * @param[in] dof_handler hp::DoFHandler object
+     * @param[in] predicates vector of predicates defining the subdomains.
+     * <code>@p predicates[i]</code> returns true for a cell if it
+     * belongs to subdomain with index i.
+     * @param[in] predicate_colors vector of colors (unsigned int) associated
+     * with each subdomain.
+     * @param[out] cellwise_color_predicate_map A map of maps used to associate
+     * predicate colors in cells with predicate ids.
+     * @param[out] fe_sets a vector of color lists
+     */
+    template <int dim, int spacedim>
+    void
+    set_cellwise_color_set_and_fe_index(
+      hp::DoFHandler<dim, spacedim> &                       dof_handler,
+      const std::vector<predicate_function<dim, spacedim>> &predicates,
+      const std::vector<unsigned int> &                     predicate_colors,
+      std::map<unsigned int, std::map<unsigned int, unsigned int>>
+        &                                  cellwise_color_predicate_map,
+      std::vector<std::set<unsigned int>> &fe_sets);
+
+    /**
+     * A function that returns a vector of enrichment functions corresponding
+     * to a color. The size of the vector is equal to total number of different
+     * colors associated with predicates (i.e subdomains).
+     *
+     * Assume that a cell has a active predicates with ids 4 (color = 1) and
+     * 5 (color = 2). cellwise_color_predicate_map has this information
+     * provided we know the material id.
+     *
+     * Now a call to color_enrichment[1](cell) should in turn call
```

Comment:
The constructed `color_enrichments` is such that
`color_enrichments[color](cell)` will return return a pointer to function with id=4, i.e. enrichments[4].
In other words, using the previously collected information in this function we translate a vector of user provided enrichment functions into a vector of functions suitable for FE_Enriched class.

Reply:
@niveshd i think you missed this one.